Add opm migration to drop unused tables

esoergel opened the pull request:
@benrudolph @czue 
https://github.com/dimagi/commcare-hq/pull/8809#discussion_r43047357

esoergel committed:
Add opm migration to drop unused tables

esoergel commented on custom/opm/migrations/0001_drop_old_fluff_tables.py:
```diff
@@ -0,0 +1,34 @@
+        'fluff_OpmHealthStatusAllInfoFluff',
+        'fluff_VhndAvailabilityFluff',
+    ]:
+        Table(table_name, metadata).drop()
```
you can pass `check_first=True` to `drop` to pass if the table isn't found, but I figured it's better for that to fail hard.

czue replied:
don't think it's better to fail hard (since anyone who installs HQ after this migration was added should never have those tables). Also, this is why the build is failing.

esoergel replied:
Oh right - I guess whatever created the tables no longer exists, so this isn't undoing those.

esoergel replied:
Okay, updated that.  I also reset the log level to its previous value after this function completes - I noticed this is chatty.

czue commented:
:x: 
```
  Applying opm.0001_drop_old_fluff_tables...No handlers could be found for logger "sqlalchemy.engine.base.Engine"
Traceback (most recent call last):
  File "manage.py", line 69, in <module>
    execute_from_command_line(sys.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 161, in handle
    executor.migrate(targets, plan, fake=options.get("fake", False))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 68, in migrate
    self.apply_migration(migration, fake=fake)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 102, in apply_migration
    migration.apply(project_state, schema_editor)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/migrations/migration.py", line 108, in apply
    operation.database_forwards(self.app_label, schema_editor, project_state, new_state)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/migrations/operations/special.py", line 117, in database_forwards
    self.code(from_state.render(), schema_editor)
  File "/home/travis/build/dimagi/commcare-hq/custom/opm/migrations/0001_drop_old_fluff_tables.py", line 24, in drop_tables
    Table(table_name, metadata).drop()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/schema.py", line 628, in drop
    checkfirst=checkfirst)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1479, in _run_visitor
    conn._run_visitor(visitorcallable, element, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1122, in _run_visitor
    **kwargs).traverse_single(element)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/sql/visitors.py", line 111, in traverse_single
    return meth(obj, **kw)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/ddl.py", line 184, in visit_table
    self.connection.execute(schema.DropTable(table))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 662, in execute
    params)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 720, in _execute_ddl
    compiled
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 874, in _execute_context
    context)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1024, in _handle_dbapi_exception
    exc_info
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 195, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 867, in _execute_context
    context)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 324, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (ProgrammingError) table "fluff_OPMHierarchyFluff" does not exist
 '\nDROP TABLE "fluff_OPMHierarchyFluff"' {}
```

esoergel committed:
Add util to temporarily alter log levels

Also backport ContextDecorator from python 3.  I saw this just the other
day and it looks like an awesome pattern, and a much clearer way to
write decorators.

esoergel committed:
Don't attempt to delete tables if they don't exist

esoergel committed:
Make sure log level gets reset afterwards

benrudolph commented on corehq/util/decorators.py:
```diff
@@ -22,3 +35,20 @@ def _handle_exceptions(*args, **kwargs):
     return _outer
+
+
+class change_log_level(ContextDecorator):
```
dig it
